lib/helpers.js

```diff
@@ -1061,14 +1008,157 @@ function matchComponentName (classString) {
   return /^[a-z0-9./_]+$/i.exec(classString);
 }
 
+/**
+ * @typedef {Object} LaunchableActivity
+ * @property {string} name
+ * @property {string} [label]
+ * @property {string} [icon]
+ */
+
+/**
+ * @typedef {Object} ApkManifest
+ * @property {string} name Package name, for example 'io.appium.android.apis'
+ * @property {number} versionCode
+ * @property {string} [versionName]
+ * @property {string} [platformBuildVersionName]
+ * @property {number} [platformBuildVersionCode]
+ * @property {number} compileSdkVersion
+ * @property {string} [compileSdkVersionCodename]
+ * @property {number} minSdkVersion
+ * @property {number} [targetSdkVersion]
+ * @property {string[]} usesPermissions List of requested permissions
+ * @property {LaunchableActivity} launchableActivity
+ * @property {string[]} locales List of supported locales
+ * @property {string[]} architectures List of supported architectures. Could be empty for older apps.
+ * @property {number[]} densities List of supported display densities
+ */
+
+/**
+ * Extracts various package manifest details
+ * from the given application file.
+ *
+ * @this {import('./adb.js').ADB}
+ * @param {string} apkPath Full path to the application file.
+ * @returns {Promise<ApkManifest>}
+ */
+export async function readPackageManifest(apkPath) {
+  await this.initAapt2();
+  const aapt2Binary = (/** @type {import('@appium/types').StringRecord} */ (this.binaries)).aapt2;
+
+  const args = ['dump', 'badging', apkPath];
+  log.debug(`Reading package manifest: '${util.quote([aapt2Binary, ...args])}'`);
+  /** @type {string} */
+  let stdout;
+  try {
+    ({stdout} = await exec(aapt2Binary, args));
+  } catch (e) {
+    throw new Error(`Cannot read the manifest from '${apkPath}'. Original error: ${e.stderr || e.message}`);
```

Comment:
The `badging` does not exist in old SDK tools like v27. The error message could also suggest to using newer SDK versions with `aapt2Binary` path info.


This is the error msg example
```
./aapt2 dump badging /path/to/apk
W/ziparchive(1423366): Unable to open 'badging': No such file or directory
badging: error: No such file or directory.
```

Reply:
Added custom error message